spec/lib/msf/base/sessions/command_shell_spec.rb

```diff
@@ -0,0 +1,159 @@
+
+    it 'should have the correct type' do
+      expect(subject.type).to eq(type)
+    end
+    it 'should have the correct description' do
+      expect(subject.desc).to eq(description)
+    end
+
+    it 'should not support aborting the process running in the session' do
+      expect(subject.abort_foreground_supported).to be(true)
+    end
+
+    it 'should initialise the shell by default' do
+      expect(subject.shell_init).to be(true)
+    end
+
+    describe 'Builtin commands' do
+      COMMANDS.each do |command|
+        next if command == :help
+
+        describe "#cmd_#{command}" do
+          context 'when called with the `-h` argument' do
+            it 'should call the corresponding help function' do
+              expect(subject).to receive("cmd_#{command}_help")
+              subject.send("cmd_#{command}", '-h')
+            end
+          end
+
+          context 'when called with the `--help` argument' do
+            it 'should call the corresponding help function' do
+              expect(subject).to receive("cmd_#{command}_help")
+              subject.send("cmd_#{command}", '--help')
+            end
+          end
+        end
+      end
+    end
+
+    describe '#run_builtin_cmd' do
+      COMMANDS.each do |command|
+        context "when called with `#{command}`" do
+          it "should call cmd_#{command}" do
+            expect(subject).to receive("cmd_#{command}")
+            subject.run_builtin_cmd(command.to_s, nil)
+          end
+        end
+      end
+    end
+
+    describe '#run_single' do
+      COMMANDS.each do |command|
+        context "when called with builtin command `#{command}`" do
+          it 'should call the builtin function' do
+            expect(subject).to receive(:run_builtin_cmd)
+            subject.run_single(command.to_s)
+          end
+        end
+      end
+
+      context 'when called with a non-builtin command' do
+        let(:cmd) { 'some_command' }
+        it 'should write the command to the shell' do
+          expect(subject).to receive(:shell_write).with("#{cmd}\n")
+          subject.run_single(cmd)
+        end
+      end
+    end
+
+    describe '#process_autoruns' do
+      let(:initial_auto_run_script) { 'initial_auto_run_script' }
+      let(:auto_run_script) { 'auto_run_script' }
+
+      context 'The datastore is empty' do
+        let(:datastore) do
+          Msf::DataStore.new
+        end
+        it 'should not execute any script' do
+          is_expected.not_to receive(:execute_script)
+          subject.process_autoruns(datastore)
+        end
+      end
+
+      context 'The datastore contains an `InitialAutoRunScript`' do
+        let(:datastore) do
+          datastore = Msf::DataStore.new
+          datastore['InitialAutoRunScript'] = initial_auto_run_script
+          datastore
+        end
+        it 'should execute the script' do
+          is_expected.to receive(:execute_script).with(initial_auto_run_script)
```

Comment:
Not a blocker; expect to receive is an antipattern as it doesn't follow the normal pattern of arrange/act/assert

https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/

i.e. prefer adding a `expect(...).to have_received(...)` after the 'act', instead of `expect(...). to receive(...)` before your 'act'